designs/documents.md

```diff
@@ -0,0 +1,424 @@
+the `schema` keyword-only argument.
+
+### Accessors
+
+The `Document` type has accessors for each Python data type that is generated
+from Smithy. This means, for example, that there is an `as_integer`, but not an
+`as_short` or `as_big_decimal`. If it is necessary to distinguish between the
+different integer or float sizes, the `shape_type` can be used.
+
+When any of the accessor methods are called, the underlying data is asserted to
+be of the requested type before it is returned, raising an exception if it is
+not. In some cases these types are coerced where they're compatible. For
+example, attempting to access a `Decimal` from a document that contains a
+`float` will return a `Decimal` representation of that float.
+
+The following example shows how accessors may be used:
+
+```python
+# With no or limited prior knowledge, a match can be used to handle different
+# possible types.
+match document.shape_type:
+    case ShapeType.STRING:
+        print(document.as_string())
+    case ShapeType.BLOB:
+        print(document.as_blob().decode("utf-8"))
+    case _:
+        pass
+
+# With concrete prior knowledge, the accessor can be called directly.
+# If the underlying type doesn't match, an exception will be raised.
+print(f"Total: {document.as_float()}")
+```
+
+Note that the values of lists and maps are `Document`s themselves.
+
+#### DocumentValue
+
+In addition to accessing Smithy data types, `Document`s will have a method to
+return the document as a plain Python object without any context. These
+`DocumentValue`s are similar to a JSON value, except that they include
+`datetime.datetime`, `bytes`, and `Decimal` as possible types. `Doucment`s may
+also be construced by providing a `DocumentValue`. Other implementations of the
+`Document` type may not include this functionality, but for Python it is a huge
+ease-of-use improvement for those that don't need the extra saftey and
+functionality of the full `Document` type.
+
+Currently the base `Document` implementation lazily converts between the two
+representations and caches the results. This is done for the sake of speed, but
+has an unfortunate space cost. This behavior is not part of the API, however,
+and MAY be changed without backwards compatibility implications.
+
+```python
+>>> document = Document({"foo": "bar"})
+>>> document.shape_type
+ShapeType.DOCUMENT
+>>> document.as_value()
+{"foo": "bar"}
+>>> document["foo"]
+Document(value="bar")
+```
+
+#### Structures and Unions
+
+Smithy structures and unions are generated into their own individual classes
+in Python, and so it isn't possible to have a unique accessor for each of them.
+Instead, `Document`s have a generic `as_shape` method that supports converting
+the `Document` to any `DeserialzeableShape`.
+
+Similarly, a `from_shape` method allows for converting any `SerializeableShape`
+into a document. When a `Document` is constructed this way, it will retain the
```

Comment:
What is the expected behavior for when the input value does not fit into the specified shape?  I'm assuming that will raise an error based on the contents here, but I'd like to see that behavior recorded in this doc.